<launch>
  <!-- Arguments -->
  <arg name="model" default="waffle_pi"/>
  <arg name="map_file" default="$(find turtlebot3_navigation)/maps/map.yaml"/>
  <arg name="move_forward_only" default="false"/>
  <arg name="cmd_vel_topic" default="/cmd_vel"/>
  <arg name="odom_topic" default="odom"/>
  <arg name="param_set" default="confident"/>

  <!-- Turtlebot3 -->
  <include file="$(find turtlebot3_bringup)/launch/turtlebot3_remote.launch">
    <arg name="model" value="$(arg model)"/>
  </include>

  <!-- Gazebo -->
  <include file="$(find turtlebot3_gazebo)/launch/turtlebot3_world.launch"/>

  <!-- Map server -->
  <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)"/>

  <!-- AMCL -->
  <include file="$(find turtlebot3_navigation)/launch/amcl.launch">
    <arg name="initial_pose_x" default="-2.0"/>
    <arg name="initial_pose_y" default="-0.5"/>
    <arg name="initial_pose_a" default="0.0"/>
  </include>

  <!-- AMCL path drawing -->
  <node name="amcl_path_node" pkg="rrt_planner" type="amcl_with_path.py" output="screen">
    <param name="use_sim_time" value="true" />
  </node>

  <!-- move_base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS" />
    <param name="base_global_planner" value="rrt_planner/RRTPlannerROS"/>
    <rosparam file="$(find rrt_planner)/config/rrt_planner.yaml" command="load"/>

    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/costmap_common_params_$(arg model).yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/costmap_common_params_$(arg model).yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/move_base_params.yaml" command="load" />
    <rosparam file="$(find rrt_planner)/param/$(arg param_set)/dwa_local_planner_params_$(arg model).yaml" command="load" />

    <param name="DWAPlannerROS/min_vel_x" value="0.0" if="$(arg move_forward_only)"/>
    <param name="planner_frequency" value="0.0" />

    <remap from="cmd_vel" to="$(arg cmd_vel_topic)"/>
    <remap from="odom" to="$(arg odom_topic)"/>
  </node>

  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz" required="true" args="-d $(find rrt_planner)/rviz/rrt_turtlebot_sim.rviz"/>

  <!-- Goal Publisher (Python script node) -->
  <node pkg="rrt_planner" type="goal_publisher.py" name="goal_publisher" output="screen">
    <!-- Goals take the form of [x, y, Î¸] or [x, y, qz, qw] -->
    <param name="goal_list" value="[
      (2.0, -0.5, 1.57080),
      (2.0, 0.5, 3.14159),
      (-2.0, 0.5, -1.57080),
      (-2.0, -0.5, 0.0)
    ]" />
    <param name="start_delay" value="0" />  
  </node>
</launch>
